### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Jyumonji-x**
yaml的config还没有实现，其他的基本都试着写了一下

### CODE REVIEW

File: src/ModelInterface.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    def load_model(self, name):
+        try:
+            Model = getattr(import_module(
+                'model.' + name), name)
+        except:
+            raise ValueError(
+                f'Invalid Module File Name {name}!')
+        self.model = Model()
+
+    def configure_optimizers(self):
```
Comment (rexxy-sasori) [inline src/ModelInterface.py:32]: This is ok, but there are so many other optimizers. Sometimes I even want to implement my own optimizer. How do I handle the case where my optimizer is not SGD? Also you are hardcoding only the learning rate. What about weight decay? momentum? Finally, what if I want to have my own learning rate schedule?

File: src/datamodel/CIFAR10DataModel.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            self.train, self.val = random_split(cifar10_full, [45000, 5000])
+
+        # Assign test dataset for use in dataloader(s)
+        if stage == "test" or stage is None:
+            self.test = datasets.CIFAR10(self.data_dir, train=False, transform=self.transform)
+
+        if stage == "predict" or stage is None:
+            self.predict = datasets.CIFAR10(self.data_dir, train=False, transform=self.transform)
+
+    def train_dataloader(self):
```
Comment (rexxy-sasori) [inline src/datamodel/CIFAR10DataModel.py:29]: You are mostly right for the data module case, but what if I want to pass in other arguments like num_workers, drop_last, shuffle, sampler?

File: src/datamodel/CIFAR10DataModel.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+class CIFAR10DataModule(pl.LightningDataModule):
+    def __init__(self, data_dir: str = "./data"):
+        super().__init__()
+        self.data_dir = data_dir
+        self.transform = transforms.ToTensor()
+
+    def prepare_data(self):
+        # download
+        datasets.CIFAR10('data', train=True, download=True, transform=self.transform)
```
Comment (rexxy-sasori) [inline src/datamodel/CIFAR10DataModel.py]: You are hardcoding the directory of the dataset. Please make it a variable.

File: src/ModelInterface.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return self.model.forward(x)
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        prediction = self(x)
+        loss = nn.CrossEntropyLoss()(prediction, y)
+        acc = torchmetrics.functional.accuracy(prediction, y)
+        self.log("training_loss", loss)
+        self.log("training_acc", acc)
+        self.log("learning-rate", self.lr)
```
Comment (rexxy-sasori) [inline src/ModelInterface.py]: You don't need to log you learning rate here. Please remove this line and use LearningRateMonitor from the callback module

File: src/ModelInterface.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        super().__init__()
+        self.save_hyperparameters()
+        self.load_model(model_name)
+        self.lr = 1e-2
+
+    def load_model(self, name):
+        try:
+            Model = getattr(import_module(
+                'model.' + name), name)
+        except:
```
Comment (rexxy-sasori) [inline src/ModelInterface.py:27]: This is a bad practice that you are not specifying exactly what error you are expecting. Just pointing this out. No change needed here

File: src/main.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+if __name__ == '__main__':
+
+    ckpt_callback = pl.callbacks.ModelCheckpoint(
+        monitor='val_loss',
+        save_top_k=1,
+        mode='min'
+    )
+
+    class EpochCallback(Callback):
```
Comment (rexxy-sasori) [inline src/main.py:32]: Good job! You found it.

File: src/main.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+        def on_epoch_end(self, trainer, pl_module):
+            epoch_time = time.time() - pl_module.epochTimeKeeper
+            pl_module.log("epoch_time", epoch_time)
+
+    logger = CSVLogger("logs", name="resnet32_cifar")
+    trainer = pl.Trainer(max_epochs=5, gpus=0, callbacks=[ckpt_callback, EpochCallback()],
+                         logger=logger,
+                         max_time="00:01:00:00", check_val_every_n_epoch=1,
+                         log_every_n_steps=1, strategy="ddp")
```
Comment (rexxy-sasori) [inline src/main.py]: Again hardcoding.

File: src/model/ResNet32.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                                                          0))
+
+    def forward(self, x):
+        out = nn.functional.relu(self.bn1(self.conv1(x)))
+        out = self.bn2(self.conv2(out))
+        out += self.shortcut(x)
+        out = nn.functional.relu(out)
+        return out
+
+class ResNet32(nn.Module):
```
Comment (rexxy-sasori) [inline src/model/ResNet32.py:35]: Specifying the depth here is Ok, but later on we will make the depth a variable as well, since there is ResNet-56 or ResNet152

File: src/main.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    class EpochCallback(Callback):
+
+        def on_epoch_start(self, trainer, pl_module):
+            pl_module.epochTimeKeeper = time.time()
+
+        def on_epoch_end(self, trainer, pl_module):
+            epoch_time = time.time() - pl_module.epochTimeKeeper
+            pl_module.log("epoch_time", epoch_time)
+
+    logger = CSVLogger("logs", name="resnet32_cifar")
```
Comment (rexxy-sasori) [inline src/main.py]: Also hardcoding. Please make it a variable.

**Review by rexxy-sasori — CHANGES_REQUESTED** [review]
Nice job of using lightning for the first time. I pointed out a couple places that you need to change. We will move on to the next phase after I finish reviewing your updated version. Looking forward to it.

File: src/main.py [hunk 10]
```diff
@@ -1,13 +1,28 @@
...
+    gpus = 0
+    max_time = "00:01:00:00"
+    check_val_every_n_epoch = 1
+    log_every_n_steps = 1
+    strategy = "ddp"
+    optimizer = optim.SGD
+    lr = 1e-2
+    model_name = "ResNet32"
+
+    lr_monitor = pl.callbacks.LearningRateMonitor(logging_interval='step')
```
Comment (Jyumonji-x) [inline src/main.py:25]: I used variables instead. I have no idea whether it's the right way to modify the hardcode.

File: src/ModelInterface.py [hunk 11]
```diff
@@ -14,12 +13,13 @@
 
 
 class ModelInterface(pl.LightningModule):
-    def __init__(self, model_name):
+    def __init__(self, model_name, optimizer_func, lr):
```
Comment (Jyumonji-x) [inline src/ModelInterface.py:16]: I finally used a variable. But I didn't find some ways to enable learning rate schedulers.

File: src/datamodel/CIFAR10DataModel.py (hunk 2 above)
Comment (Jyumonji-x) [reply src/datamodel/CIFAR10DataModel.py:29]: I tried to solve the problem by using dictionary variables, but I don't know whether it works well.

File: src/ModelInterface.py (hunk 1 above)
Comment (Jyumonji-x) [reply src/ModelInterface.py:32]: I used 'optimizer_func' to enable changes in optimizers. However, I found that I can't add more arguments such as weight decay. I tried to just use instantiated optimizer, but I couldn't provide the parameters.
For learning rate, I finally used a variable, but I didn't find some ways to enable learning rate schedulers.

**Review by Jyumonji-x — COMMENTED** [review]
I partly finished the modification of hardcode, but I have no idea in some of the problems.
